### ISSUE TITLE
DataParallel spills memory to GPU #0

### ISSUE THREAD

**Issue by deepbrain** [pytorch/pytorch#7071]
## Issue description

on a multi GPU system, if a GPU != 0 is used, pytorch will still allocate some memory on the GPU 0 - see the nvidia-smi screenshot below.

## Code example

```python 
import torch
import torch.nn as nn
from torch.autograd import Variable
import torch.optim as optim........

torch.backends.cudnn.benchmark = True

class TestNet(nn.Module):
    def __init__(self):
        super(TestNet, self).__init__()
        self.net1 = nn.Linear(100, 200)
        self.net2 = nn.Linear(200, 1)
        self.bn = nn.BatchNorm1d(200)
        self.sigmoid = nn.Sigmoid()
        self.ReLU = nn.ReLU(inplace=False)
        self.drop = nn.Dropout(0.5)
...............
    def forward(self, V):
        return self.sigmoid(self.net2(self.drop(self.ReLU(self.bn(self.net1(V)))))).squeeze().


GPU = 1
net = TestNet()
criterion = nn.BCELoss()
net.cuda(GPU)
criterion.cuda(GPU)
V = Variable(torch.randn(100, 100)).cuda(GPU)
label = Variable(torch.randn(100)).cuda(GPU)
optim_betas = (0.9, 0.999)
optimizer = optim.Adam(filter(lambda p: p.requires_grad, net.parameters()), lr=0.01, betas=optim_betas)

net.train()
for i in range(0,1000000):
    net.zero_grad()
    r = net(V)
    err = criterion(r, label)
    err.backward()...

```
tester@NEWTITAN:~$ nvidia-smi
Sat Apr 28 23:45:03 2018       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 387.26                 Driver Version: 387.26                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  TITAN X (Pascal)    Off  | 00000000:05:00.0 Off |                  N/A |
| 23%   33C    P2    57W / 250W |    401MiB / 12189MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  GeForce GTX 108...  Off  | 00000000:06:00.0 Off |                  N/A |
| 29%   33C    P2    58W / 250W |    549MiB / 11172MiB |      6%      Default |
+-------------------------------+----------------------+----------------------+
|   2  GeForce GTX 108...  Off  | 00000000:09:00.0 Off |                  N/A |
| 29%   36C    P8    10W / 250W |     10MiB / 11172MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0      1262      C   python                                       391MiB |
|    1      1262      C   python                                       539MiB |
+-----------------------------------------------------------------------------+

**Comment by WenmuZhou** [comment]
I encountered the same problem

**Comment by wenqiangzh** [comment]
I met the same problem, tooooo. So how to deal with it?

**Comment by zou3519** [comment]
Is this fixed by https://github.com/pytorch/pytorch/pull/7392 ?

**Comment by zou3519** [comment]
Verified that the bug has been fixed on master:
```
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    1   3987668      C   /home/rzou/local/miniconda3/bin/python       362MiB |
+-----------------------------------------------------------------------------+

```

**Comment by kostyanoob** [comment]
I had this problem even after the bug-fix. You should be able to debug your code and detect the line that causes the memory spill. (by debugging step by step and watching the nvidia-smi tool for memory increase in gpu0). Then you should wrap the problematic line with a scope. 

For example, I found that the loss computation was spilling to GPU0, whereas I wanted to use GPU3 only, so this is what I did:
```
with torch.cuda.device(3):
    loss = hinge_loss(output, y)
```
